### PR TITLE
Nickytonline/homepage main feed component

### PR DESCRIPTION
**Pull request by nickytonline**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This is the main feed component for the hompeage makeover.

[Live Storybook](https://devto-storybook.now.sh/) with some additions for this PR

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

LEGO caveman to suave LEGO man in a tuxedo

### CODE REVIEW

File: app/javascript/articles/components/CommentsCount.jsx [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import { h } from 'preact';
 import PropTypes from 'prop-types';
 
-export const CommentsCount = ({ count, articlePath, icon }) => {
+export const CommentsCount = ({ count, articlePath, icon, className }) => {
   if (count > 0) {
     return (
-      <div className="article-engagement-count comments-count">
+      <div
+        className={`article-engagement-count comments-count ${className || ''}`}
```
Comment (nickytonline) [inline app/javascript/articles/components/CommentsCount.jsx]: This is to support the `<FeaturedArticle />` component which takes additional "featured" classes.

File: app/javascript/articles/components/TagList.jsx [hunk 2]
```diff
@@ -1,8 +1,9 @@
 import { h } from 'preact';
+import PropTypes from 'prop-types';
 import { tagPropTypes } from '../../src/components/common-prop-types';
 
-export const TagList = ({ tags = [] }) => (
-  <div className="tags">
+export const TagList = ({ tags = [], className }) => (
+  <div className={`tags ${className || ''}`}>
```
Comment (nickytonline) [inline app/javascript/articles/components/TagList.jsx]: This is to support the `<FeaturedArticle />` component which takes additional "featured" classes.

File: app/views/articles/index.html.erb [hunk 3]
```diff
@@ -82,73 +83,11 @@
...
-
-      <div class="substories" id="substories">
-        <% if @stories.any? %>
-          <%= render "stories/main_stories_feed" %>
-        <% end %>
-      </div>
+      <div id="homepage-feed"></div>
 
+      <%# TODO REMOVE LOADING DIV AND ACCOMPANYING JS %>
       <div class="loading-articles" id="loading-articles">
```
Comment (nickytonline) [inline app/views/articles/index.html.erb]: I didn't remove this piece of markup yet, or maybe I won't because I think it's used on other pages like search.

File: app/javascript/articles/Article.jsx [hunk 4]
```diff
@@ -99,7 +99,7 @@ export const Article = ({
         </a>
       </h4>
 
-      <TagList tags={article.tag_list || article.cached_tag_list_array} />
+      <TagList tags={article.tag_list} />
```
Comment (nickytonline) [inline app/javascript/articles/Article.jsx]: @joshpuetz's work on the feed eliminates the need for checking for `cached_tag_list_array`. It is always just `tag_list` now.

File: app/javascript/articles/FeaturedArticle.jsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+      >
+        <div
+          className="single-article big-article"
+          data-content-user-id={article.user_id}
+        >
+          <div
+            className="picture image-final"
+            style={{
+              backgroundColor: article.main_image_background_hex_color,
+              backgroundImage: `url(${cloud_cover_url(article.main_image)})`,
```
Comment (nickytonline) [inline app/javascript/articles/FeaturedArticle.jsx]: This is the `cloud_cover_url` function I was talking about @benhalpern. I'm faking it here on the client-side by simply returning the `main_image` URL, but I imagine the `cloud_cover_url` server-side function returns a CDN URL?

File: app/javascript/packs/Feed.jsx.erb [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import { h, Component } from 'preact';
+import PropTypes from 'prop-types';
+
+<% helpers = ActionController::Base.helpers %>
+
+const FEED_ICONS = {
+  COMMENTS_ICON: '<%= helpers.image_path("comments-bubble.png") %>',
```
Comment (nickytonline) [inline app/javascript/packs/Feed.jsx.erb]: This needs to be an ERB template so that we can generate Fastly URLs for these icons, when in PROD.

File: app/javascript/packs/homePage.jsx [hunk 7]
```diff
@@ -1,7 +1,19 @@
 import { h, render } from 'preact';
+import { renderFeed } from './homePageFeed';
 
 /* global userData */
 
+// This logic is similar to that in initScrolling.js.erb
+// that prevents the classic Algolia scrolling for the front page.
+const frontPageFeedPathNames = new Map([
```
Comment (nickytonline) [inline app/javascript/packs/homePage.jsx]: This is the best way for now that I could come up with for updating the feed URL when we grab article data.

File: app/javascript/packs/homePage.jsx [hunk 8]
```diff
@@ -54,17 +66,20 @@ function renderTagsFollowed(tagsFollowedContainer, user = userData()) {
...
   if (userStatus === 'logged-out') {
-    // Nothing to do, the user is not logged on.
+    renderFeed(feedTimeFrame);
     return;
   }
 
   if (userStatus === 'logged-in' && user !== null) {
     // We have user data, render followed tags.
     clearTimeout(waitingForDataLoad);
+    renderFeed(feedTimeFrame);
```
Comment (nickytonline) [inline app/javascript/packs/homePage.jsx]: We render the feed regardless of whether the user is logged on or not, but in the case of a logged on user, we want to ensure the user data has been set before making the call to the feed endpoint.

File: app/javascript/packs/homePage.jsx [hunk 9]
```diff
@@ -73,7 +88,18 @@ let waitingForDataLoad = setTimeout(function dataLoadedCheck() {
   waitingForDataLoad = setTimeout(dataLoadedCheck, 40);
 }, 40);
 
-InstantClick.on('receive', (_url, body, title) => {
+InstantClick.on('change', () => {
+  const url = new URL(window.location);
```
Comment (nickytonline) [inline app/javascript/packs/homePage.jsx]: I originally had the feed rendering in the `InstantClick` on `receive` event, but this caused unintended behaviour. It did indeed work, but since the feed time frame prop of the `<Feed />` component changed, the current time frame tab would get updated with the time frame data of the time frame tab that was being hovered on.

**Comment by nickytonline** [comment]
@joshpuetz @benhalpern. This is still a Draft PR, but the feed works on the homepage now, if you just want to take it for a spin.

Things to do still are load podcast episodes and to one either implement the Algolia paging within the `<Feed />` component or to leave it as is.

Leaving Algolia Scrolling as-is:
PRO:
* Works like it did before

CON
* Will not use the new Preact components. This is not necessarily a deal-breaker, as other parts of the site will still be doing this as well, but it'd be nice to have it all Preactified.

**Comment by joshpuetz** [comment]
I'm definitely team "leave Algolia in place" for now: cleaning that up can be it's own separate PR/cycle!

File: app/javascript/articles/FeaturedArticle.jsx (hunk 5 above)
Comment (benhalpern) [reply app/javascript/articles/FeaturedArticle.jsx]: Ah yes. @joshpuetz I think `main_image` should deliver the fully CDN-ified version, e.g. `cloud_cover_url` which had been working via view helper...

Similar to `cached_user`/`user` we can let the server worry about sending the appropriate version.

Comment (nickytonline) [reply app/javascript/articles/FeaturedArticle.jsx]: We would need to return the cloud URL in the feed for the first story, i.e. featured story. I can't do it at build time like the other ERB/webpacker stuff, because the feature story changes.

Comment (nickytonline) [reply app/javascript/articles/FeaturedArticle.jsx]: Ooops, responded on stale comments. 🙃

Comment (nickytonline) [reply app/javascript/articles/FeaturedArticle.jsx]: So I'll assume the proper URL is being sent for the main image then.

Comment (joshpuetz) [reply app/javascript/articles/FeaturedArticle.jsx]: Just checked the feed endpoint, and we're sending what I think are non-cdn-ified locations for images (ex: https://dev-to-uploads.s3.amazonaws.com/i/9bmcx1wglljih84nx65u.png). Will adjust in the endpoint @nickytonline!

File: app/assets/javascripts/initializers/initializeBaseTracking.js.erb [hunk 10]
```diff
@@ -15,7 +15,6 @@ function initializeBaseTracking() {
       ga('set', 'anonymizeIp', true);
       ga('send', 'pageview', location.pathname + location.search);
       clearInterval(waitingOnGA);
-      logImpressions();
```
Comment (nickytonline) [inline app/assets/javascripts/initializers/initializeBaseTracking.js.erb]: @benhalpern, I've reworked GA for the featured story.

This does not work with the new `<Feed />` component. It gets called before the element with the `'featured-story-marker'` ID exists. So now this is handled in the homePageFeed.jsx.

File: app/javascript/packs/homePageFeed.jsx [hunk 11]
```diff
@@ -3,6 +3,29 @@ import { Article, LoadingArticle } from '../articles';
...
 import { Feed } from './Feed.jsx.erb';
 
+/**
+ * Sends analytics about the featured article.
+ *
+ * @param {number} articleId
+ */
+function sendFeaturedArticleAnalytics(articleId) {
+  (function logFeaturedArticleImpression() {
+    if (!window.ga || !ga.create) {
```
Comment (nickytonline) [inline app/javascript/packs/homePageFeed.jsx]: This should be defined already, but...

File: app/javascript/packs/homePageFeed.jsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+ * Renders the main feed.
+ */
+export const renderFeed = timeFrame => {
+  const feedContainer = document.getElementById('homepage-feed');
+
+  render(
+    <Feed
+      timeFrame={timeFrame}
+      renderFeed={({ feedItems, feedIcons }) => {
+        if (feedItems.length === 0) {
```
Comment (nickytonline) [inline app/javascript/packs/homePageFeed.jsx]: We're never going to have an empty <Feed /> since there is so much content on the site, so realistically, when there are no articles, it's just the component loading up and waiting for data from the feed endpoint.

**Comment by nickytonline** [comment]
TODO: I also have to fix some Capybara tests.

File: app/assets/javascripts/initializers/initScrolling.js.erb [hunk 13]
```diff
@@ -152,7 +152,8 @@ function insertArticles(articles) {
   if (el) {el.outerHTML = "";}
   articles.forEach(function insertAnArticle(article) {
     var existingEl = document.getElementById("article-link-"+article.id);
-    if ( existingEl &&
+    if ( !['/', '/top/week', '/top/month', '/top/year', '/top/infinity', '/latest'].includes(window.location.pathname) &&
```
Comment (nickytonline) [inline app/assets/javascripts/initializers/initScrolling.js.erb]: This prevents articles from being overwritten with potentially newer Algolia data. As discussed @benhalpern, in the case of the home page, this is not necessary since the new feed endpoint has up to feed data already.

**Comment by nickytonline** [comment]
Taking this out of Draft mode @benhalpern @joshpuetz since Ben is taking this over.

The main `<Feed />` component is all done. I've modified it so that it only loads the new feed for logged on users.

Stuff to be aware of/left to do:

[ ] Ensure stuff in `initializeFetchFollowedArticles()` still works
[ ] Pretty sure `insertTimes()` is no longer necessary. It’s called from `initializeFetchFollowedArticles()` 
[ ] Revert ERB so that logged out users still see server-side rendered feed. I’ve done all the work to load the <Feed /> component only for a logged on user. 
[ ] I used mock data to load today’s podcasts into the feed, but we’ll need to test for real podcast episodes loading in feed.
[ ] Some Capybara tests are failing

File: app/assets/javascripts/initializers/initializeBaseTracking.js.erb [hunk 14]
```diff
@@ -67,14 +66,6 @@ function eventListening(){
   }
 }
 
-function logImpressions() {
```
Comment (nickytonline) [inline app/assets/javascripts/initializers/initializeBaseTracking.js.erb]: This probably needs to go back in for logged out users since as discussed, logged out users will be using server-side rendering.

File: app/assets/javascripts/initializers/initializeReadingListIcons.js [hunk 15]
```diff
@@ -14,7 +12,9 @@ function initializeReadingListIcons() {
 
 // set SAVE or SAVED articles buttons
 function setReadingListButtonsState() {
-  var readingListButtons = document.getElementsByClassName('bookmark-button');
+  var readingListButtons = document.querySelectorAll(
+    '.bookmark-button:not([data-initial-feed])',
```
Comment (nickytonline) [inline app/assets/javascripts/initializers/initializeReadingListIcons.js]: I added `:not([data-initial-feed]` which is on the `<SaveButton />` component so that this only picks up the reading list buttons from Algolia paging.

File: app/javascript/articles/Feed.jsx [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+import { FEED_ICONS } from '../packs/feedIcons.js.erb';
+
+/* global userData sendHapticMessage showModal buttonFormData renderNewSidebarCount */
+
+export class Feed extends Component {
+  componentDidMount() {
+    const { timeFrame } = this.props;
+    const { reading_list_ids = [] } = userData(); // eslint-disable-line camelcase
+
+    this.setState({ bookmarkedFeedItems: new Set(reading_list_ids) });
```
Comment (nickytonline) [inline app/javascript/articles/Feed.jsx]: I went with a `Set` because the current code was an array of IDs which meant that the operation to find out if an article is in the reading list was n<sup>2<sup>. Now it is log n.

File: app/javascript/articles/Feed.jsx [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    getCsrfToken()
+      .then(sendFetch('reaction-creation', data))
+      // eslint-disable-next-line consistent-return
+      .then(response => {
+        if (response.status === 200) {
+          return response.json().then(json => {
+            const articleId = Number(button.dataset.reactableId);
+
+            this.setState(previousState => {
+              const { bookmarkedFeedItems } = previousState;
```
Comment (nickytonline) [inline app/javascript/articles/Feed.jsx]: We need the previous state as it contains the other buttons that were saved unsaved.

File: app/javascript/articles/components/CommentsCount.jsx [hunk 18]
```diff
@@ -1,10 +1,14 @@
...
 import PropTypes from 'prop-types';
 
-export const CommentsCount = ({ count, articlePath, icon }) => {
+export const CommentsCount = ({ count, articlePath, icon, className }) => {
   if (count > 0) {
     return (
-      <div className="article-engagement-count comments-count">
+      <div
+        className={`article-engagement-count comments-count${
+          className ? ` ${className}` : ''
```
Comment (nickytonline) [inline app/javascript/articles/components/CommentsCount.jsx]: This change is for the `<FeaturedArticle />` which takes additional CSS classes.

File: app/javascript/articles/components/SaveButton.jsx [hunk 19]
```diff
@@ -1,42 +1,68 @@
...
-        type="button"
-        style={{ width: '122px' }}
-        className="article-engagement-count engage-button follow-action-button"
-        data-info={`{"id":${article.id},"className":"User"}`}
-        data-follow-action-button
-      >
-        &nbsp;
-      </button>
-    );
+export class SaveButton extends Component {
```
Comment (nickytonline) [inline app/javascript/articles/components/SaveButton.jsx]: Converted this to a stateful component to manage button text.

File: app/javascript/articles/components/TagList.jsx [hunk 20]
```diff
@@ -1,8 +1,9 @@
 import { h } from 'preact';
+import PropTypes from 'prop-types';
 import { tagPropTypes } from '../../src/components/common-prop-types';
 
-export const TagList = ({ tags = [] }) => (
-  <div className="tags">
+export const TagList = ({ tags = [], className }) => (
+  <div className={`tags${className ? ` ${className}` : ''}`}>
```
Comment (nickytonline) [inline app/javascript/articles/components/TagList.jsx]: This change is for the `<FeaturedArticle />` which takes additional CSS classes.

File: app/views/articles/index.html.erb [hunk 21]
```diff
@@ -82,73 +83,11 @@
           <img src="<%= asset_path "lightning.svg" %>" alt="right-sidebar-nav">
         </button>
       </div>
+      <%# END Feed menu bar %>
 
-      <% if @featured_story.id %>
```
Comment (nickytonline) [inline app/views/articles/index.html.erb]: As discussed yesterday @benhalpern, we`ll need to revert this change and only render this section if they are logged out.

Comment (benhalpern) [reply app/views/articles/index.html.erb]: Done

**Comment by benhalpern** [comment]
I think this is ready or close to ready.

File: app/javascript/articles/Feed.jsx [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    Feed.getFeedItems(timeFrame).then(feedItems => {
+
+      // Ensure first article is one with a main_image
+      const featuredStory = feedItems.find(story => story.main_image !== null);
+      // Remove that first one from the array.
+      const index = feedItems.indexOf(featuredStory);
+      feedItems.splice(index, 1);
+      const subStories = feedItems;   
+      const organizedFeedItems = [featuredStory, subStories].flat();  
+      console.log(organizedFeedItems);
```
Comment (rhymes) [inline app/javascript/articles/Feed.jsx]: noticed this as I was testing checking to see if there were any errors in the console
